Refactor Braze Banner code

tjmw opened the pull request:
## What does this change?

Following on from #25628 and #25636 continue refactoring the Braze Banner code to make it possible to onboard another slot on frontend powered by Braze (header notifications).

The main changes here are:

* Instead of `brazeBanner.ts` knowing how to get a `BrazeMessages` instance, we create the instance outside and pass it in. This will allow us to use the instance elsewhere.
* We've refactored the logic around checking the required dependencies before loading the Braze SDK from DCR, which has worked nicely there.
* General refactoring of functions out of `brazeBanner.js`.

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

## Screenshots

Test Braze Banner showing up successfully in CODE:

<img width="1426" alt="Screenshot 2022-11-10 at 11 27 35" src="https://user-images.githubusercontent.com/379839/201083001-2b8016c4-458d-4ff3-b5dc-222f3d545e34.png">

## What is the value of this and can you measure success?

Refactoring to make onboarding of a new slot on the page powered by Braze possible.

## Checklist

### Tested

- [ ] Locally
- [x] On CODE (optional)

KaliedaRik reviewed (approved):
Refactor looks good

tjmw commented:
I paired with @arelra and we were satisfied that the Commercial E2E failure isn't related to this work, so I'm going to go ahead and merge (with a fix coming for the E2E tests soon).

mxdvl commented on static/src/javascripts/projects/common/modules/commercial/braze/brazeMessageInterface.ts:
```diff
@@ -0,0 +1,14 @@
+// Note on BrazeMessageInterface vs BrazeMessage:
+// BrazeMessage is the actual class which we wrap messages in returned from the
+// Braze SDK (in @guardian/braze-components). BrazeMessageInterface represents
+// the public interface supplied by instances of that class. In the message
+// forcing logic it's hard to create and return a BrazeMessage from the forced
+// json. A better approach would probably be to change the code in
+// @guardian/braze-components to return a BrazeMessageInterface (which an
+// instance of BrazeMessage would conform to) and at least the type definitions
+// would all live in one centralised place.
```
You could use JSDoc here if you wished!

mxdvl commented on static/src/javascripts/projects/common/modules/commercial/braze/getMessageFromUrlFragment.ts:
```diff
@@ -0,0 +1,65 @@
+import { log } from '@guardian/libs';
+import type { BrazeMessageInterface } from './brazeMessageInterface';
+
+const FORCE_BRAZE_ALLOWLIST = [
+	'preview.gutools.co.uk',
+	'preview.code.dev-gutools.co.uk',
+	'localhost',
+	'm.thegulocal.com',
+];
```
Using an `as const` assertion can help, here. It will also prevent anyone from mutating the array.

mxdvl commented on static/src/javascripts/projects/common/modules/commercial/braze/getMessageFromUrlFragment.ts:
```diff
@@ -0,0 +1,65 @@
+	'm.thegulocal.com',
+];
+
+const isExtrasData = (data: unknown): data is Record<string, string> => {
+	if (typeof data === 'object' && data != null) {
+		return Object.values(data).every((value) => typeof value === 'string');
+	}
+	return false;
+};
```
@guardian/libs has an `isObject` and `isString` method, which you may want to use for consistency.

mxdvl commented on static/src/javascripts/projects/common/modules/commercial/braze/getMessageFromUrlFragment.ts:
```diff
@@ -0,0 +1,65 @@
+				return;
+			}
+
+			const forcedMessage = hashString.slice(
+				hashString.indexOf(`${key}=`) + key.length + 1,
+				hashString.length,
+			);
```
If these are formatted like search paeans, `URLSearchParams` could help simplify a bit.

mxdvl reviewed (commented):
Major refactor, very nice!

Added some small thoughts for your consideration.

tjmw commented:
Thanks for the review @mxdvl, we've got more work to do in this area so will pick up addressing your comments in future PRs.

> How come this does not need to be reproduced in DCR?

We're refactoring frontend to support multiple slots powered by Braze. Right now Braze on frontend is only integrated with the sticky bottom banner slot but on DCR it already supports multiple slots - banner, epic and header notifications (behind a feature flag).
